### ISSUE TITLE
Hard crash in Godot 3.3 with large particle emitter and large image

### ISSUE THREAD

**Issue by benbishopnz** [godotengine/godot#47732]
**Godot version:**
Exists in 3.3 RC8, and as far back as 3.2.4 beta1.
3.2.3 stable does not crash.

**OS/device including version:**
Windows 10, GLES3 project

**Issue description:**
While testing the new 3.3 RC8 build with my game, one of the levels in my game crashes while playing. The crash timing is inconsistent (ie. sometimes it takes a few seconds to happen, sometimes a minute or more - but it has happened every time I test it.)

The game crashes without logging any errors. I checked the Monitors tab and ran the Profiler and couldn't see anything out of the ordinary going on either.

The main difference between this level and the others in my game is there is a large particle emitter in the background (with many long-lived particles) and a large image (3840x1080) scrolling in the background.

I intend to test removing these elements later today and test again. If you have any other ideas of things to test or more information I can gather, please ask below.

**Steps to reproduce:**
Here's a video of it happening:
https://youtu.be/d2qgUlWLf1I

Not much to see unfortunately. The stars are particles I mentioned.

**Minimal reproduction project:**
TBA, might not be possible in this case.

**Comment by benbishopnz** [comment]
BVH doesn't seem to affect it but I managed to get through the entire level while disabling Batching.
I also tried deleting the Particles2D node and the Sprite node with the large background and that seemed to work too, but I got a crash if either still existed.

**Comment by akien-mga** [comment]
CC @lawnjelly as it seems it might be related to GLES3 batching.

To debug further we'll really need either a reproduction project, or more debug info that you could extract from your project (compile Godot from source in debug mode, get a crash backtrace in Visual Studio or similar debugging tool, etc.).

Sounds like extracting your problematic Particle2D and Sprite nodes to a separate project might be a potential way to get a minimal reproduction project you could share?

**Comment by lawnjelly** [comment]
Agree will probably need a reproduction project I suspect, or for you to run a debug build. If you can run from a console / terminal that might also give you more debugging info if an error is being output.

Large textures have caused problems in the past but it is usually with hardware that flat out doesn't support them, and this doesn't appear to be in your case. Interestingly particles are not batched. They do go through the same pipeline however, so there could be some interaction (like a state which is not being reset).

You could also turn on `rendering/batching/debug/diagnose_frame` and post the log here, but I'm not convinced it will pin it down.

The other option is to email Akien or myself the project privately if you don't want it made public.

**Comment by benbishopnz** [comment]
Thanks guys. I'm having difficulty reproducing it in a minimal project I'm
afraid, so I will try to get the diagnose_frame log first. I don't have any
experience in compiling Godot myself yet, but it might be the time to learn
how :)

**Comment by benbishopnz** [comment]
Turning `max_join_item_commands` also fixed the problem.
Replacing the large image didn't seem to do anything. There are quite a lot of moving parts and transparent images being drawn on top of each other, for what it's worth (although, no more than other levels that I haven't seen this crash happen...)

Here's a `diagnose_frame` log I captured with an early crash:
```--- Debugging process started ---
Godot Engine v3.3.rc8.official - https://godotengine.org
OpenGL ES 3.0 Renderer: GeForce GTX 970/PCIe/SSE2
OpenGL ES Batching: ON
 
canvas_begin FRAME 520
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 192] {255 255 255 255 }
        joined_item 1 refs,
                        batch D 0-1 PA
        joined_item 6 refs,
                        batch R 0-1 [0 - 210] {255 255 255 255 }
                        batch R 0-1 [1 - 211] {255 255 255 255 }
                        batch R 0-1 [2 - 216] {255 255 255 255 }
                        batch R 0-1 [3 - 218] {255 255 255 255 }
                        batch R 0-5 [4 - 686] {119 119 119 255 } MULTI color
                        batch R 0-8 [5 - 685] {255 255 255 255 } MULTI color
        joined_item 1 refs,
                        batch R 0-8 [0 - 685] {41 138 204 255 } MULTI
        joined_item 4 refs,
                        batch R 0-17 [0 - 685] {255 255 255 255 } MULTI
                        batch R 0-1 [1 - 228] {255 255 255 255 }
canvas_end

canvas_begin FRAME 1044
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 950] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 950] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 976] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 976] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 944] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 944] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 945] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 945] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 947] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 947] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 946] {255 255 255 255 }
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 946] {255 255 255 255 }
items
items
        joined_item 1 refs,
                (z 60)
                        batch D 0-1 PA
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 2578] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch D 0-1 PA
        joined_item 3 refs,
                (z 60)
                        batch R 0-3 [0 - 2588] {255 255 255 255 } MULTI
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 989] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 700] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 2547] {255 255 255 255 }
items
items
items
        joined_item 2 refs,  ( prevent_color_baking, use_modulate_fvf,  )
                        batch R 0-2 [0 - 943] {497 298 388 255 } MULTI
items
        joined_item 2 refs,  ( prevent_color_baking, use_modulate_fvf,  )
                        batch R 0-2 [0 - 943] {497 298 388 255 } MULTI
items
canvas_end

canvas_begin FRAME 1044
items
canvas_end

canvas_begin FRAME 1044
items
        joined_item 23 refs,
                        batch R 0-5 [0 - 2599] {0 0 0 150 } MULTI
                        batch R 5-5 [0 - 2599] {255 255 255 255 } MULTI color
                        batch R 10-2 [0 - 2599] {0 0 0 150 } MULTI color
                        batch R 12-2 [0 - 2599] {255 255 255 255 } MULTI color
                        batch R 0-6 [1 - 2600] {0 0 0 255 } MULTI color
                        batch R 6-6 [1 - 2600] {255 255 255 255 } MULTI color
                        batch R 0-4 [1 - 2600] {0 0 0 255 } MULTI color
                        batch R 4-4 [1 - 2600] {255 255 255 255 } MULTI color
                        batch R 0-1 [2 - 2565] {255 255 255 255 }
                        batch R 0-1 [3 - 963] {255 255 255 255 }
                        batch R 0-1 [4 - 2570] {255 255 255 255 }
                        batch R 0-1 [3 - 963] {255 255 255 255 }
                        batch R 0-5 [5 - 2611] {0 0 0 255 } MULTI color
                        batch R 5-5 [5 - 2611] {255 255 255 255 } MULTI color
                        batch R 0-6 [6 - 2612] {0 0 0 255 } MULTI color
                        batch R 6-6 [6 - 2612] {255 255 255 255 } MULTI color
                        batch R 0-4 [7 - 2613] {0 0 0 255 } MULTI color
                        batch R 4-4 [7 - 2613] {255 249 172 255 } MULTI color
                        batch R 0-7 [5 - 2611] {0 0 0 255 } MULTI color
                        batch R 7-7 [5 - 2611] {255 255 255 255 } MULTI color
                        batch R 0-3 [7 - 2613] {0 0 0 255 } MULTI color
                        batch R 3-3 [7 - 2613] {255 255 255 255 } MULTI color
                        batch R 0-6 [5 - 2611] {0 0 0 255 } MULTI color
                        batch R 6-6 [5 - 2611] {255 255 255 255 } MULTI color
                        batch R 0-1 [8 - 954] {255 255 255 255 }
                        batch R 0-3 [9 - 955] {255 255 255 255 } MULTI
                        batch R 0-1 [10 - 958] {255 255 255 255 }
                        batch R 0-1 [11 - 959] {255 255 255 255 }
                        batch R 0-6 [5 - 2611] {0 0 0 255 } MULTI color
                        batch R 6-6 [5 - 2611] {255 255 255 255 } MULTI color
                        batch R 0-1 [12 - 960] {255 255 255 255 }
                        batch R 1-1 [13 - 962] {255 255 255 255 }
                        batch R 2-1 [14 - 961] {255 255 255 255 }
                        batch R 0-5 [5 - 2611] {0 0 0 255 } MULTI color
                        batch R 5-5 [5 - 2611] {255 255 255 255 } MULTI color
                        batch R 0-1 [15 - 160] {255 255 255 255 }
canvas_end

canvas_begin FRAME 1044
items
        joined_item 2 refs,
                        batch R 0-1 [0 - 161] {255 255 255 255 }
                        batch R 0-1 [1 - 163] {255 255 255 255 }
canvas_end

canvas_begin FRAME 1611
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 2807] {255 255 255 255 }
        joined_item 1 refs,
                        batch D 0-1 PA
        joined_item 4 refs,
                        batch R 0-1 [0 - 2825] {255 255 255 255 }
                        batch R 0-1 [1 - 2826] {255 255 255 255 }
                        batch R 0-1 [2 - 2830] {255 255 255 255 }
                        batch R 0-5 [3 - 3279] {119 119 119 255 } MULTI color
canvas_end

canvas_begin FRAME 2117
items
        joined_item 1 refs,
                        batch R 0-1 [0 - 3569] {255 255 255 255 }
        joined_item 1 refs,
                        batch D 0-1 PA
        joined_item 6 refs,
                        batch R 0-6 [0 - 4642] {255 255 255 255 } MULTI
        joined_item 1 refs,
                        batch R 1-1 [0 - 3309] {0 82 255 85 }
                        batch R 3-1 [1 - 4977] {255 255 255 255 } color
                        batch R 5-1 [0 - 3309] {0 82 255 85 } color
                        batch R 7-1 [2 - 4979] {255 255 255 255 } color
                        batch R 9-1 [0 - 3309] {0 82 255 85 } color
                        batch R 11-1 [3 - 4981] {255 255 255 255 } color
                        batch R 13-1 [0 - 3309] {0 82 255 85 } color
                        batch R 15-1 [4 - 4983] {255 255 255 255 } color
                        batch R 17-1 [0 - 3309] {0 82 255 85 } color
                        batch R 19-1 [5 - 4985] {255 255 255 255 } color
                        batch R 21-1 [0 - 3309] {0 82 255 85 } color
                        batch R 23-1 [6 - 4987] {255 255 255 255 } color
                        batch R 25-1 [0 - 3309] {0 82 255 85 } color
                        batch R 27-1 [7 - 4989] {255 255 255 255 } color
                        batch R 29-1 [0 - 3309] {0 82 255 85 } color
                        batch R 31-1 [8 - 4991] {255 255 255 255 } color
                        batch R 33-1 [0 - 3309] {0 82 255 85 } color
                        batch R 35-1 [9 - 5007] {255 255 255 255 } color
                        batch R 37-1 [0 - 3309] {0 82 255 85 } color
                        batch R 39-1 [10 - 5009] {255 255 255 255 } color
                        batch R 41-1 [0 - 3309] {0 82 255 85 } color
                        batch R 43-1 [11 - 5013] {255 255 255 255 } color
                        batch R 45-1 [0 - 3309] {0 82 255 85 } color
                        batch R 47-1 [12 - 5015] {255 255 255 255 } color
                        batch R 49-1 [0 - 3309] {0 82 255 85 } color
                        batch R 51-1 [13 - 5017] {255 255 255 255 } color
                        batch R 53-1 [0 - 3309] {0 82 255 85 } color
                        batch R 55-1 [14 - 5019] {255 255 255 255 } color
                        batch R 57-1 [0 - 3309] {0 82 255 85 } color
                        batch R 59-1 [15 - 5021] {255 255 255 255 } color
                        batch R 61-1 [0 - 3309] {0 82 255 85 } color
                        batch R 63-1 [16 - 5023] {255 255 255 255 } color
                        batch R 65-1 [0 - 3309] {0 82 255 85 } color
                        batch R 67-1 [17 - 5025] {255 255 255 255 } color
                        batch R 69-1 [0 - 3309] {0 82 255 85 } color
                        batch R 71-1 [18 - 5027] {255 255 255 255 } color
                        batch R 73-1 [0 - 3309] {0 82 255 85 } color
                        batch R 75-1 [19 - 5029] {255 255 255 255 } color
                        batch R 77-1 [0 - 3309] {0 82 255 85 } color
                        batch R 79-1 [20 - 5031] {255 255 255 255 } color
                        batch R 81-1 [0 - 3309] {0 82 255 85 } color
                        batch R 83-1 [21 - 5033] {255 255 255 255 } color
                        batch R 85-1 [0 - 3309] {0 82 255 85 } color
                        batch R 87-1 [22 - 5035] {255 255 255 255 } color
                        batch R 89-1 [0 - 3309] {0 82 255 85 } color
                        batch R 91-1 [23 - 5037] {255 255 255 255 } color
                        batch R 93-1 [0 - 3309] {0 82 255 85 } color
                        batch R 95-1 [24 - 5039] {255 255 255 255 } color
        joined_item 2 refs,
                (z 50)
                        batch R 0-1 [0 - 4752] {255 255 255 255 }
                        batch R 0-1 [1 - 4753] {255 255 255 255 }
        joined_item 1 refs,
                (z 50)
                        batch D 0-1 PA
        joined_item 1 refs,
                (z 50)
                        batch R 0-1 [0 - 4727] {255 255 255 255 }
        joined_item 1 refs,
                (z 50)
                        batch D 0-1 PA
        joined_item 2 refs,
                (z 50)
                        batch R 0-1 [0 - 4727] {185 185 185 255 }
                        batch R 0-1 [1 - 4743] {228 228 228 255 } color
        joined_item 1 refs,
                (z 59)
                        batch P 0-6 {255 255 255 255 } MULTI
        joined_item 1 refs,
                (z 59)
                        batch R 0-1 [0 - 3303] {255 255 255 255 }
        joined_item 1 refs,
                (z 59)
                        batch R 0-1 [0 - 4743] {255 255 255 255 }
        joined_item 1 refs,
                (z 59)
                        batch P 0-6 {255 255 255 255 } MULTI
        joined_item 1 refs,
                (z 59)
                        batch R 0-1 [0 - 3303] {255 255 255 255 }
        joined_item 6 refs,
                (z 60)
                        batch R 0-1 [0 - 3431] {255 255 255 255 }
                        batch R 0-1 [1 - 3432] {255 255 255 255 }
                        batch R 0-1 [2 - 3433] {255 255 255 255 }
                        batch R 0-1 [0 - 3431] {255 255 255 255 }
                        batch R 0-1 [1 - 3432] {255 255 255 255 }
                        batch R 0-1 [2 - 3433] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch D 0-1 PA
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 4732] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch D 0-1 PA
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 3608] {255 255 255 255 }
        joined_item 1 refs,
                (z 60)
                        batch R 0-1 [0 - 3303] {255 255 255 255 }
        joined_item 2 refs,
                (z 60)
                        batch R 0-1 [0 - 4626] {255 255 255 255 }
                        batch R 0-1 [1 - 4743] {292 292 292 255 } color
        joined_item 1 refs,
                (z 61)
                        batch P 0-6 {255 255 255 255 } MULTI
        joined_item 1 refs,
                (z 61)
                        batch R 0-1 [0 - 3303] {255 255 255 255 }
canvas_end

canvas_begin FRAME 2117
items
canvas_end

canvas_begin FRAME 2117
items
        joined_item 21 refs,
                        batch R 0-5 [0 - 4766] {0 0 0 150 } MULTI
                        batch R 5-5 [0 - 4766] {255 255 255 255 } MULTI color
                        batch R 10-1 [0 - 4766] {0 0 0 150 } color
                        batch R 11-1 [0 - 4766] {255 255 255 255 } color
                        batch R 0-6 [1 - 4767] {0 0 0 255 } MULTI color
                        batch R 6-6 [1 - 4767] {255 255 255 255 } MULTI color
                        batch R 0-4 [1 - 4767] {0 0 0 255 } MULTI color
                        batch R 4-4 [1 - 4767] {255 255 255 255 } MULTI color
                        batch R 0-1 [2 - 3623] {255 255 255 255 }
                        batch R 0-1 [3 - 3583] {255 255 255 255 }
                        batch R 0-1 [4 - 4653] {255 255 255 255 }
                        batch R 0-1 [3 - 3583] {255 255 255 255 }
                        batch R 0-5 [5 - 4778] {0 0 0 255 } MULTI color
                        batch R 5-5 [5 - 4778] {255 255 255 255 } MULTI color
                        batch R 0-6 [6 - 4779] {0 0 0 255 } MULTI color
                        batch R 6-6 [6 - 4779] {255 255 255 255 } MULTI color
                        batch R 0-4 [7 - 4780] {0 0 0 255 } MULTI color
                        batch R 4-4 [7 - 4780] {255 255 255 255 } MULTI color
                        batch R 0-7 [5 - 4778] {0 0 0 255 } MULTI color
                        batch R 7-7 [5 - 4778] {255 255 255 255 } MULTI color
                        batch R 0-2 [7 - 4780] {0 0 0 255 } MULTI color
                        batch R 2-2 [7 - 4780] {255 255 255 255 } MULTI color
                        batch R 0-6 [5 - 4778] {0 0 0 255 } MULTI color
                        batch R 6-6 [5 - 4778] {255 255 255 255 } MULTI color
                        batch R 0-1 [8 - 3574] {255 255 255 255 }
                        batch R 0-2 [9 - 3575] {255 255 255 255 } MULTI
                        batch R 0-1 [10 - 3578] {255 255 255 255 }
                        batch R 0-1 [11 - 3579] {255 255 255 255 }
                        batch R 0-5 [5 - 4778] {0 0 0 255 } MULTI color
                        batch R 5-5 [5 - 4778] {255 255 255 255 } MULTI color
                        batch R 0-2 [12 - 160] {255 255 255 255 } MULTI
canvas_end

**Comment by lawnjelly** [comment]
Thanks for posting the log, I can't see anything immediately jumping out as causing the problem so it still would need you to run a debug build to get a stack trace or a min reproduction project. Turning off joining fixing it is useful info, it may be some particular combination of primitives you are using that there is a bug handling.

Another possibility - you aren't by any chance doing anything with threads that might affect the render items? That can be problematic.

**Comment by benbishopnz** [comment]
No threading. I'll attempt to get a stack trace and get back to you soon.

**Comment by benbishopnz** [comment]
Alright, I've got Visual Studio building/debugging 3.3.x branch but am not sure exactly what I need to do to get a stack trace out of it. It's late so I am going to head to bed but can try further tomorrow. If this isn't a good forum to work through it I can provide an email address or we can chat on Discord / IRC or whatever.

**Comment by Calinou** [comment]
You can get a crash stack trace simply by running a Godot binary that contains debug symbols using a command prompt. Once Godot crashes, a stack trace will be printed. This works on all desktop platforms.

**Comment by benbishopnz** [comment]
Alright, hopefully this is what you wanted:

```
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] RasterizerCanvasGLES3::render_batches (godot-3.x\drivers\gles3\rasterizer_canvas_gles3.cpp:534)
[1] RasterizerCanvasBatcher<RasterizerCanvasGLES3,RasterizerStorageGLES3>::flush_render_batches (godot-3.x\drivers\gles_common\rasterizer_canvas_batcher.h:2459)
[2] RasterizerCanvasBatcher<RasterizerCanvasGLES3,RasterizerStorageGLES3>::render_joined_item_commands (godot-3.x\drivers\gles_common\rasterizer_canvas_batcher.h:2548)
[3] RasterizerCanvasGLES3::render_joined_item (godot-3.x\drivers\gles3\rasterizer_canvas_gles3.cpp:1488)
[4] RasterizerCanvasGLES3::canvas_render_items_implementation (godot-3.x\drivers\gles3\rasterizer_canvas_gles3.cpp:1976)
[5] RasterizerCanvasBatcher<RasterizerCanvasGLES3,RasterizerStorageGLES3>::batch_canvas_render_items_end (godot-3.x\drivers\gles_common\rasterizer_canvas_batcher.h:790)
[6] RasterizerCanvasGLES3::canvas_render_items_end (godot-3.x\drivers\gles3\rasterizer_canvas_gles3.cpp:62)
[7] VisualServerCanvas::render_canvas (godot-3.x\servers\visual\visual_server_canvas.cpp:278)
[8] VisualServerViewport::_draw_viewport (godot-3.x\servers\visual\visual_server_viewport.cpp:246)
[9] VisualServerViewport::draw_viewports (godot-3.x\servers\visual\visual_server_viewport.cpp:350)
[10] VisualServerRaster::draw (godot-3.x\servers\visual\visual_server_raster.cpp:109)
[11] VisualServerWrapMT::draw (godot-3.x\servers\visual\visual_server_wrap_mt.cpp:104)
[12] Main::iteration (godot-3.x\main\main.cpp:2139)
[13] OS_Windows::run (godot-3.x\platform\windows\os_windows.cpp:3440)
[14] widechar_main (godot-3.x\platform\windows\godot_windows.cpp:162)
[15] _main (godot-3.x\platform\windows\godot_windows.cpp:184)
[16] main (godot-3.x\platform\windows\godot_windows.cpp:196)
[17] __scrt_common_main_seh (d:\A01\_work\12\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[18] BaseThreadInitThunk
-- END OF BACKTRACE --
```

**Comment by lawnjelly** [comment]
Assuming you are compiling a current version of 3.x from github, that seems to line up to this line:

```
for (int i = batch.first_command; i < end_command; i++) {
	Item::Command *c = p_commands[i];
	switch (c->type) { // CRASH HERE
```
It would be good if you could double check this corresponds to line 534 of rasterizer_canvas_gles3.cpp in your source code (and if not post the lines around and indicate the line of 534 with an arrow or something).

We've never had a crash here before so I'm slightly suspecting your source is a different version, but if it is crashing here, it is likely to be a threading bug imo (something else modifying the render commands while the render is in progress).

**Comment by akien-mga** [comment]
This would likely prevent the crash (here at least) and replace it with an error, possibly discarding the whole frame and leading to temporarily broken rendering?

```diff
diff --git a/drivers/gles3/rasterizer_canvas_gles3.cpp b/drivers/gles3/rasterizer_canvas_gles3.cpp
index 2e7aab2ad8..1e96584640 100644
--- a/drivers/gles3/rasterizer_canvas_gles3.cpp
+++ b/drivers/gles3/rasterizer_canvas_gles3.cpp
@@ -530,6 +530,7 @@ void RasterizerCanvasGLES3::render_batches(Item::Command *const *p_commands, Ite
                                for (int i = batch.first_command; i < end_command; i++) {
 
                                        Item::Command *c = p_commands[i];
+                                       ERR_FAIL_NULL(c);
 
                                        switch (c->type) {
 
```

**Comment by lawnjelly** [comment]
Yes, but if this is happening at all, it's a big no no.

Assuming there isn't a bug in the recording of the command ids, changing the render commands while a render is in progress is breaking the contract, and it is likely to go horribly wrong with unpredictable results. So this would need to be fixed at source.

But this is all speculative until we get the OP to confirm this is the line on their version of the source code.

**Comment by benbishopnz** [comment]
Yes, you're correct @lawnjelly - the crash is happening at the line containing `switch (c->type) {`.

**Comment by Zireael07** [comment]
@benbishopnz: Can you mail the project to one of the contributors? I know it has been offered as an option for people who didn't want to publicize their commercial projects...

**Comment by benbishopnz** [comment]
Can do. Who wants it?

**Comment by lawnjelly** [comment]
I've got a copy of the project now and am debugging and can reproduce it. :+1: 
I'm now suspecting it's a rare combination of items that should be blocked from joining, which is easier to solve than a race condition.

**Comment by akien-mga** [comment]
Fixed by #47832.
